x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -237,4 +392,293 @@ public void onFailure(Exception e) {
+                    searchSource.pointInTimeBuilder(pointInTime);
+                    pointInTime.setKeepAlive(keepAlive);
+                    final SearchRequest searchRequest = new SearchRequest();
+                    searchRequest.source(searchSource);
+                    if (searchAfter != null) {
+                        searchSource.searchAfter(searchAfter);
+                    }
+                    clientWithOrigin.execute(SearchAction.INSTANCE, searchRequest, new ActionListener<>() {
+                        @Override
+                        public void onResponse(SearchResponse response) {
+                            maintenanceTask.total.compareAndSet(0L, response.getHits().getTotalHits().value);
+                            maintenanceTask.searchResponse = response;
+                            maintenanceTask.searchAfter = null;
+                            executeNext(maintenanceTask);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            complete(maintenanceTask, e);
+                        }
+                    });
+                    return;
+                }
+
+                final SearchHit[] searchHits = searchResponse.getHits().getHits();
+                if (searchHits != null && searchHits.length > 0) {
+                    final ClusterState state = clusterService.state();
+                    final BulkRequest bulkRequest = new BulkRequest();
+
+                    final TimeValue retention = periodicTaskRetention;
+                    final Instant expirationTime = Instant.ofEpochMilli(threadPool.absoluteTimeInMillis())
+                        .minus(retention.duration(), retention.timeUnit().toChronoUnit());
+
+                    // compute the list of existing searchable snapshots once
+                    Map<String, Set<String>> knownSnapshots = existingSnapshots;
+                    if (knownSnapshots == null) {
+                        knownSnapshots = listSearchableSnapshots(state);
+                        existingSnapshots = knownSnapshots;
+                    }
+
+                    final Set<String> knownRepositories = state.metadata()
+                        .custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY)
+                        .repositories()
+                        .stream()
+                        .map(RepositoryMetadata::name)
+                        .collect(Collectors.toSet());
+
+                    Object[] lastSortValues = null;
+                    for (SearchHit searchHit : searchHits) {
+                        lastSortValues = searchHit.getSortValues();
+                        assert searchHit.getId() != null;
+                        assert searchHit.hasSource();
+
+                        try {
+                            final CachedBlob cachedBlob = CachedBlob.fromSource(searchHit.getSourceAsMap());
```

Comment:
Can we add a static method to extract the created time? In this variant, we do more work, including decoding the blob.

Maybe we can use the fields API to avoid retrieving the source, I would think the timestamp can be retrieved as a doc-value?

Reply:
This is a good idea, I pushed this as part of 15ca8cdd9394fc5e1c9d2c9abec84525a64c4b13